### PR TITLE
Prop-Types Support

### PR DESCRIPTION
**Pull request by thegoldenmule**
This PR adds a very basic [prop-types](https://www.npmjs.com/package/prop-types) renderer.

```typescript
interface PrimitiveObject {
    string: string;
    bool: boolean;
    float: number;
    union: string | boolean | number;
    object: object
}

interface PrimitiveArrayObject {
    strings: string[];
    bools: boolean[];
    floats: number[];
    unions: string[] | boolean[] | number[];
    objects: object[];
}

interface CompositeObject {
    primitive: PrimitiveObject;
    primitiveArrays: PrimitiveArrayObject;
}
```

Emits:

```javascript
// For use with proptypes.

import PropTypes from 'prop-types';

export const CompositeObject = PropTypes.shape({
    primitive: PrimitiveObject,
    primitiveArrays: PrimitiveArrayObject,
});

export const PrimitiveObject = PropTypes.shape({
    bool: PropTypes.bool.isRequired,
    float: PropTypes.number.isRequired,
    object: PropTypes.object.isRequired,
    string: PropTypes.string.isRequired,
    union: PropTypes.oneOfType([PropTypes.bool,PropTypes.number,PropTypes.string]).isRequired,
});

export const PrimitiveArrayObject = PropTypes.shape({
    bools: PropTypes.arrayOf(PropTypes.bool).isRequired,
    floats: PropTypes.arrayOf(PropTypes.number).isRequired,
    objects: PropTypes.arrayOf(PropTypes.object).isRequired,
    strings: PropTypes.arrayOf(PropTypes.string).isRequired,
    unions: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.bool,PropTypes.number,PropTypes.string])).isRequired,
});

```

**Motivation**: The prop-types library is an opensource runtime type checker used in React development (among loads of other things). Having an emitter for prop-types means that my react app can validate and IDEs can autocomplete.

**Details**: I couldn't find any docs about contributing or writing tests, so I modeled it after the JavaScriptRenderer.

**HALP**: How should this be tested? Prop-types isn't like other structures that can just be serialized/deserialized. I have some ideas for custom tests, but was unsure where to put them.

### CODE REVIEW

**Comment by thegoldenmule** [comment]
BTW this build is failing because there is no space left, not because of the code:

`write /usr/lib/go-1.6/pkg/linux_amd64/net/http.a: no space left on device`

**Comment by schani** [comment]
Failing Lint checks now.

**Comment by thegoldenmule** [comment]
@schani What lint checks are you referring to? I'm new to BuildKite and I don't see a specific step for linting. Happy to fix whatever the issue is, however.

**Comment by schani** [comment]
Here's a basic test scaffolding that just does a `wc` on the output:

https://github.com/quicktype/quicktype/commit/407bc7e8ad1e03ec31430ceea83a936a5dcea0fe

**Comment by schani** [comment]
Is the buildkite output hidden for you?  I thought we made it so everybody can see it.

You can run lint on your own machine with `npm run tslint`:

```
Step 54/54 : RUN npm run tslint
  | ---> Running in 5cc78d4fec48
  |  
  | > quicktype@15.0.0 tslint /app
  | > tslint --project src/cli --exclude 'src/__tests__/**' --exclude 'src/cli/get-stream/**'
  |  
  | no-unused-variable is deprecated. Since TypeScript 2.9. Please use the built-in compiler checks instead.
  |  
  | ERROR: /app/src/quicktype-core/index.ts:87:140 - Missing semicolon
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:86:32 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:87:26 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:88:29 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:89:28 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:90:28 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:92:33 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:93:25 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:97:62 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:97:70 - Missing semicolon
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:100:24 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:114:23 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:114:52 - Missing semicolon
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:125:23 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:125:63 - Missing semicolon
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:132:27 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:132:50 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:132:73 - Missing semicolon
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:137:25 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:139:25 - ' should be "
  | ERROR: /app/src/quicktype-core/language/JavaScriptPropTypes.ts:142:27 - ' should be "
 ```

**Comment by thegoldenmule** [comment]
The output is public, I was just expecting a specific build step and wasn't digging through the logs. Easily seen now, thanks.

File: test/fixtures/javascript-prop-types/main.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import fs from "fs";
+import process from "process";
+import * as TopLevel from "./toplevel.js";
+
+const sample = process.argv[2];
+const json = fs.readFileSync(sample);
```
Comment (schani) [inline test/fixtures/javascript-prop-types/main.js:6]: It doesn't seem like this is necessary, is it?  All you want is for `toplevel.js` to be imported correctly.  You don't have to copy the same file, then, either.

I'm fine to merge this if you don't want to change it anymore.

Comment (thegoldenmule) [reply test/fixtures/javascript-prop-types/main.js:6]: Yah, this is in progress. I was actually planning on having it read the json and validate it against the input prop-types.

**Pull request merged**